cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -423,8 +427,13 @@ public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
                                     @Nullable Set<PluginClass> additionalPlugins,
                                     Map<String, String> properties) throws Exception {
     // check that the current user has write privileges on the namespace
-    authorizerInstantiatorService.get().enforce(artifactId.getNamespace().toEntityId(),
-                                                SecurityRequestContext.toPrincipal(), Action.WRITE);
+    // Skip authorization enforcement for system user. This is so the SystemArtifactLoader thread can add system
+    // artifacts.
+    Principal principal = SecurityRequestContext.toPrincipal();
+    NamespaceId namespace = artifactId.getNamespace().toEntityId();
+    if (!(Principal.SYSTEM.equals(principal) && NamespaceId.SYSTEM.equals(namespace))) {
```

Comment:
I can move it to a different method. I need to think about the other two use cases you mentioned. I'll likely address them in a different PR though, since without this particular fix the default namespace does not get created and system artifacts do not get loaded.